### PR TITLE
etcdmain: proxy does not need to belong to the discovered cluster

### LINKED ISSUE #3411: Proxy doesnt start with SRV - cannot find local etcd member "default" in SRV records
I have a 3 node cluster and 1 proxy running on node01. Proxy fails to start with error
*etcdmain:  error setting up initial cluster: cannot find local etcd member "default" in SRV records*

**Node01**
/var/lib/etcd/etcd -advertise-client-urls https://node01.etcd.dev.com:2379 -initial-cluster-state new -name node01 -data-dir /var/lib/etcd/cluster/datadir -listen-peer-urls https://node01.etcd.dev.com:2380 -initial-advertise-peer-urls https://node01.etcd.dev.com:2380 -listen-client-urls https://node01.etcd.dev.com:2379 -discovery-srv etcd.dev.com -trusted-ca-file /var/home/rm/ca.cert.pem -cert-file /var/home/rm/node01.etcd.dev.com.cert.pem -key-file /var/home/rm/node01.etcd.dev.com.pem -peer-trusted-ca-file /var/home/rm/ca.cert.pem -peer-cert-file /var/home/rm/node01.etcd.dev.com.cert.pem -peer-key-file /var/home/rm/node01.etcd.dev.com.pem -heartbeat-interval 200 -election-timeout 2000 -initial-cluster-token etcd-cluster-dev
 
**Node02**
/var/lib/etcd/etcd -advertise-client-urls https://node02.etcd.dev.com:2379 -initial-cluster-state new -name node02 -data-dir /var/lib/etcd/cluster/datadir -listen-peer-urls https://node02.etcd.dev.com:2380 -initial-advertise-peer-urls https://node02.etcd.dev.com:2380 -listen-client-urls https://node02.etcd.dev.com:2379 -discovery-srv etcd.dev.com -trusted-ca-file /var/home/rm/ca.cert.pem -cert-file /var/home/rm/node02.etcd.dev.com.cert.pem -key-file /var/home/rm/node02.etcd.dev.com.pem -peer-trusted-ca-file /var/home/rm/ca.cert.pem -peer-cert-file /var/home/rm/node02.etcd.dev.com.cert.pem -peer-key-file /var/home/rm/node02.etcd.dev.com.pem -heartbeat-interval 200 -election-timeout 2000 -initial-cluster-token etcd-cluster-dev

**Node03**
/var/lib/etcd/etcd -advertise-client-urls https://node03.etcd.dev.com:2379 -initial-cluster-state new -name node03 -data-dir /var/lib/etcd/cluster/datadir -listen-peer-urls https://node03.etcd.dev.com:2380 -initial-advertise-peer-urls https://node03.etcd.dev.com:2380 -listen-client-urls https://node03.etcd.dev.com:2379 -discovery-srv etcd.dev.com -trusted-ca-file /var/home/rm/ca.cert.pem -cert-file /var/home/rm/node03.etcd.dev.com.cert.pem -key-file /var/home/rm/node03.etcd.dev.com.pem -peer-trusted-ca-file /var/home/rm/ca.cert.pem -peer-cert-file /var/home/rm/node03.etcd.dev.com.cert.pem -peer-key-file /var/home/rm/node03.etcd.dev.com.pem -heartbeat-interval 200 -election-timeout 2000 -initial-cluster-token etcd-cluster-dev

**Proxy**
/var/lib/etcdProxy/etcd  -data-dir /var/lib/etcdProxy/proxy/datadir -listen-client-urls http://node01.etcd.dev.com:2377 -discovery-srv etcd.dev.com -trusted-ca-file /var/home/rm/ca.cert.pem -cert-file /var/home/rm/node01.etcd.dev.com.cert.pem -key-file /var/home/rm/node01.etcd.dev.com.pem -proxy on

**Proxy Logs**
2015-08-31 21:17:07.580366 I | etcdmain: etcd Version: 2.2.0-rc.0
2015-08-31 21:17:07.580457 I | etcdmain: Git SHA: dc3e027
2015-08-31 21:17:07.580479 I | etcdmain: Go Version: go1.5
2015-08-31 21:17:07.580493 I | etcdmain: Go OS/Arch: linux/amd64
2015-08-31 21:17:07.580510 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
2015-08-31 21:17:07.581346 N | discovery: got bootstrap from DNS for etcd-server-ssl at https://node01.etcd.dev.com:2380
2015-08-31 21:17:07.581386 N | discovery: got bootstrap from DNS for etcd-server-ssl at https://node02.etcd.dev.com:2380
2015-08-31 21:17:07.581417 N | discovery: got bootstrap from DNS for etcd-server-ssl at https://node03.etcd.dev.com:2380
2015-08-31 21:17:07.582634 W | discovery: discovery: error querying DNS SRV records for _etcd-server lookup _etcd-server._tcp.etcd.dev.com on 10.12.12.76:53: no such host
2015-08-31 21:17:07.582701 C | etcdmain: error setting up initial cluster: cannot find local etcd member "default" in SRV records

**Node03 logs**
2015-08-31 21:15:46.656078 I | etcdmain: etcd Version: 2.2.0-rc.0
2015-08-31 21:15:46.656147 I | etcdmain: Git SHA: dc3e027
2015-08-31 21:15:46.656157 I | etcdmain: Go Version: go1.5
2015-08-31 21:15:46.656176 I | etcdmain: Go OS/Arch: linux/amd64
2015-08-31 21:15:46.656192 I | etcdmain: setting maximum number of CPUs to 2, total number of available CPUs is 2
2015-08-31 21:15:46.656241 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2015-08-31 21:15:46.657038 N | discovery: got bootstrap from DNS for etcd-server-ssl at https://node03.etcd.dev.com:2380
2015-08-31 21:15:46.657076 N | discovery: got bootstrap from DNS for etcd-server-ssl at https://node01.etcd.dev.com:2380
2015-08-31 21:15:46.657101 N | discovery: got bootstrap from DNS for etcd-server-ssl at https://node02.etcd.dev.com:2380
2015-08-31 21:15:46.660721 W | discovery: discovery: error querying DNS SRV records for _etcd-server lookup _etcd-server._tcp.etcd.dev.com on 10.12.12.76:53: no such host
2015-08-31 21:15:46.662525 I | etcdmain: peerTLS: cert = /var/home/rm/node03.etcd.dev.com.cert.pem, key = /var/home/rm/node03.etcd.dev.com.pem, ca = , trusted-ca = /var/home/rm/ca.cert.pem, client-cert-auth = false
2015-08-31 21:15:46.664231 I | etcdmain: listening for peers on https://node03.etcd.dev.com:2380
2015-08-31 21:15:46.664263 I | etcdmain: clientTLS: cert = /var/home/rm/node03.etcd.dev.com.cert.pem, key = /var/home/rm/node03.etcd.dev.com.pem, ca = , trusted-ca = /var/home/rm/ca.cert.pem, client-cert-auth = false
2015-08-31 21:15:46.666093 I | etcdmain: listening for client requests on https://node03.etcd.dev.com:2379
2015-08-31 21:15:46.666297 I | etcdserver: name = node03
2015-08-31 21:15:46.666318 I | etcdserver: data dir = /var/lib/etcd/cluster/datadir
2015-08-31 21:15:46.666330 I | etcdserver: member dir = /var/lib/etcd/cluster/datadir/member
2015-08-31 21:15:46.666342 I | etcdserver: heartbeat = 200ms
2015-08-31 21:15:46.666350 I | etcdserver: election = 2000ms
2015-08-31 21:15:46.666362 I | etcdserver: snapshot count = 10000
2015-08-31 21:15:46.666377 I | etcdserver: advertise client URLs = https://node03.etcd.dev.com:2379
2015-08-31 21:15:46.677630 I | etcdserver: restarting member 4c5854cca2fdc9c in cluster baf099e94f57d45e at commit index 4689
2015-08-31 21:15:46.678889 I | raft: 4c5854cca2fdc9c became follower at term 224
2015-08-31 21:15:46.679011 I | raft: newRaft 4c5854cca2fdc9c [peers: [], term: 224, commit: 4689, applied: 0, lastindex: 4689, lastterm: 224]
2015-08-31 21:15:46.680961 I | etcdserver: starting server... [version: 2.2.0-rc.0, cluster version: to_be_decided]
2015-08-31 21:15:46.681902 N | etcdserver: added local member 4c5854cca2fdc9c [https://node03.etcd.dev.com:2380] to cluster baf099e94f57d45e
2015-08-31 21:15:46.685753 N | etcdserver: added member 5f97c3330e24a56d [https://node02.etcd.dev.com:2380] to cluster baf099e94f57d45e
2015-08-31 21:15:46.693508 N | etcdserver: added member f95caeae097ab105 [https://node01.etcd.dev.com:2380] to cluster baf099e94f57d45e
2015-08-31 21:15:46.693984 N | etcdserver: set the initial cluster version to 2.1
2015-08-31 21:15:46.694663 N | etcdserver: updated the cluster version from 2.1 to 2.2
2015-08-31 21:15:46.749562 I | rafthttp: the connection with 5f97c3330e24a56d became active
2015-08-31 21:15:46.758727 I | rafthttp: the connection with f95caeae097ab105 became active
2015-08-31 21:15:46.815396 I | raft: raft.node: 4c5854cca2fdc9c elected leader f95caeae097ab105 at term 224
2015-08-31 21:15:46.838640 I | etcdserver: published {Name:node03 ClientURLs:[https://node03.etcd.dev.com:2379]} to cluster baf099e94f57d45e

**etcdctl cluster-health**
./etcdctl --peers="https://node01.etcd.dev.com:2379,https://node02.etcd.dev.com:2379,https://node03.etcd.dev.com:2379" --cert-file=/var/home/rm/node01.etcd.dev.com.cert.pem --key-file=/var/home/rm/node01.etcd.dev.com.pem --ca-file=/var/home/rm/ca.cert.pem cluster-health
member 4c5854cca2fdc9c is healthy: got healthy result from https://node03.etcd.dev.com:2379
member 5f97c3330e24a56d is healthy: got healthy result from https://node02.etcd.dev.com:2379
member f95caeae097ab105 is healthy: got healthy result from https://node01.etcd.dev.com:2379
cluster is healthy

As per the [documentaion](https://github.com/coreos/etcd/blob/master/Documentation/clustering.md)
*You might see the an error like cannot find local etcd $name from SRV records.. That means the etcd member fails to find itself from the cluster defined in SRV records. The resolved address in -initial-advertise-peer-urls must match one of the resolved addresses in the SRV targets.*

I checked and it doesn't seem to be the case.

### PR DESCRIPTION
**Pull request by xiang90**
Fix #3411

### CODE REVIEW

**Comment by xiang90** [comment]
@raoofm It would be great if you can apply this patch and give it a try.

File: etcdmain/etcd.go [hunk 1]
```diff
@@ -445,8 +445,12 @@ func getPeerURLsMapAndToken(cfg *config) (urlsmap types.URLsMap, token string, e
...
 		}
 		urlsmap, err = types.NewURLsMap(clusterStr)
-		if _, ok := urlsmap[cfg.name]; !ok {
-			return nil, "", fmt.Errorf("cannot find local etcd member %q in SRV records", cfg.name)
+		// only etcd member must belong to the discovered cluster.
+		// proxy does not need to belong to the discovered cluster.
+		if which == "etcd" {
+			if _, ok := urlsmap[cfg.name]; !ok {
+				return nil, "", fmt.Errorf("cannot find local etcd member %q in SRV records", cfg.name)
+			}
```
Comment (yichengq) [inline etcdmain/etcd.go]: I think we check whether the name exists in initial-cluster later: https://github.com/coreos/etcd/blob/master/etcdserver/config.go#L94
Could we remove this check and use that one instead?

Comment (xiang90) [inline etcdmain/etcd.go]: not really. we should not check this as soon as possible. we should move discovery out of etcdserver sometime too.

Comment (yichengq) [inline etcdmain/etcd.go]: do you mean "we should check this as soon as possible" instead of "we should not check this as soon as possible"?

Comment (xiang90) [inline etcdmain/etcd.go]: we should check this as soon as possible so that we can have a better logging message for user.

**Comment by yichengq** [comment]
LGTM

**Comment by raoofm** [comment]
@xiang90 sure will test and let u know

**Comment by raoofm** [comment]
@xiang90 it works fine. Thanks for the quick response.

On a separate note, i built etcd from the master and what i noticed is the one that i built using go1.5 is more than double the size of the one that i recieved as part of v2.2.0-rc.0. Expected?
-rwxr-xr-x 1 rm rm 22158560 Sep  2 13:36 etcd
-rwxr-xr-x 1 rm rm   9992832 Sep  2 13:37 etcd.v2.2.0-rc.0

**Comment by raoofm** [comment]
@xiang90 Any comments on the size? Am I doing something wrong here

**Comment by xiang90** [comment]
@raoofm I am not sure. Will check soon.

**Pull request merged**